linera-execution/src/evm/database.rs

```diff
@@ -107,39 +96,101 @@ where
     }
 
     fn block_hash(&mut self, number: u64) -> Result<B256, ExecutionError> {
-        self.throw_error()?;
         <Self as DatabaseRef>::block_hash_ref(self, number)
     }
 }
 
 impl<Runtime> DatabaseCommit for DatabaseRuntime<Runtime>
 where
-    Runtime: ContractRuntime,
+    Runtime: BaseRuntime,
 {
-    fn commit(&mut self, changes: HashMap<Address, Account>) {
-        let result = self.commit_with_error(changes);
-        if let Err(error) = result {
-            self.commit_error = Some(error.into());
+    fn commit(&mut self, changes: EvmState) {
+        self.changes = changes;
+    }
+}
+
+impl<Runtime> DatabaseRef for DatabaseRuntime<Runtime>
+where
+    Runtime: BaseRuntime,
+{
+    type Error = ExecutionError;
+
+    fn basic_ref(&self, address: Address) -> Result<Option<AccountInfo>, ExecutionError> {
+        tracing::info!("basic_ref call at address={address:?}");
+        if !self.changes.is_empty() {
+            let account = self.changes.get(&address).unwrap();
+            return Ok(Some(account.info.clone()));
+        }
+        let mut runtime = self.runtime.lock().expect("The lock should be possible");
+        let val = self.get_contract_address_key(&address);
+        if let Some(val) = val {
+            let key_info = vec![val, KeyCategory::AccountInfo as u8];
+            let promise = runtime.read_value_bytes_new(key_info)?;
+            let result = runtime.read_value_bytes_wait(&promise)?;
+            let account_info = from_bytes_option::<AccountInfo, ViewError>(&result)?;
+            if let Some(account_info_clone) = account_info.clone() {
+                if let Some(code) = account_info_clone.code {
+                    tracing::info!("basic_ref Some: |code|={}", code.len());
+                } else {
+                    tracing::info!("basic_ref Some: |code|=None");
+                }
+            } else {
+                tracing::info!("basic_ref None");
+            }
+            Ok(account_info)
+        } else {
+            Ok(Some(AccountInfo::default()))
+        }
+    }
+
+    fn code_by_hash_ref(
+        &self,
+        _code_hash: B256,
+    ) -> Result<revm::primitives::Bytecode, ExecutionError> {
+        panic!("Functionality code_by_hash_ref not implemented");
+    }
+
+    fn storage_ref(&self, address: Address, index: U256) -> Result<U256, ExecutionError> {
+        tracing::info!("storage_ref call at address={address:?} index={index:?}");
+        if !self.changes.is_empty() {
+            let account = self.changes.get(&address).unwrap();
+            return Ok(match account.storage.get(&index) {
+                None => U256::ZERO,
+                Some(slot) => slot.present_value(),
+            });
         }
+        let val = self.get_contract_address_key(&address);
+        let Some(val) = val else {
+            panic!("There is no storage associated to Externally Owned Account");
```

Comment:
…and here. Is this all really unreachable, even with malicious EVM code?

```suggestion
            panic!("There is no storage associated to externally owned account");
```

Reply:
The address in question is the one of the running contract.
It can only be Zero or the one of the contract in question. This cannot be passed back by another application.